File: CodeSamples/kahn.R
kahn <- function(graph, n) {
  inverse <- vector("list", n)
  for (u in 1:n) {
    for (v in graph[[u]]) {
      inverse[[v]] <- c(inverse[[v]], u)
    }
  }
  stack <- which(sapply(inverse, length) == 0)
  order <- integer(0)
  while (length(stack) > 0) {
    u <- stack[length(stack)]
    stack <- stack[-length(stack)]
    order <- c(order, u)
    for (v in graph[[u]]) {
      graph[[u]] <- graph[[u]][-which(graph[[u]] == v)]
      inverse[[v]] <- inverse[[v]][-which(inverse[[v]] == u)]
      if (length(inverse[[v]]) == 0) {
        stack <- c(stack, v)
      }
    }
  }
  if (length(order) != n) {
    return("Cyclic graph")
  }
  return(order)
}

read_input <- function() {
  lines <- readLines("input.txt")
  t <- as.integer(lines[1])
  lines <- lines[-1]
  cases <- list()
  for (i in 1:t) {
    n <- as.integer(lines[1])
    lines <- lines[-1]
    graph <- list()
    while (length(lines) > 0) {
      line <- strsplit(lines[1], " ")[[1]]
      lines <- lines[-1]
      u <- as.integer(line[1])
      v <- as.integer(line[2])
      if (!exists(u, envir = graph)) {
        graph[[u]] <- list()
      }
      graph[[u]] <- c(graph[[u]], v)
    }
    cases[[i]] <- list(n = n, graph = graph)
  }
  return(list(t = t, cases = cases))
}

main <- function() {
  input <- read_input()
  output <- file("output.txt", "w")
  for (i in 1:input$t) {
    result <- kahn(input$cases[[i]]$graph, input$cases[[i]]$n)
    if (result == "Cyclic graph") {
      cat("Cyclic graph\n", file = output)
    } else {
      cat(paste(result, collapse = " "), "\n", file = output)
    }
  }
  close(output)
}

main()

File: CodeSamples/kahn.clj
(defn kahn [graph n]
  (def inverse (zipmap (range n) (repeat #{})))
  (doseq [u (keys graph)]
    (doseq [v (graph u)]
      (swap! inverse update-in [v] conj u)))

  (def stack (filter #(empty? (inverse %)) (range n)))
  (def order (atom []))

  (while (seq stack)
    (let [u (peek stack)]
      (swap! order conj u)
      (doseq [v (vec (graph u))]
        (swap! graph update-in [u] disj v)
        (swap! inverse update-in [v] disj u)
        (when (empty? (inverse v))
          (swap! stack conj v))))

  (if (= (count @order) n)
    @order
    "Cyclic graph"))

(def input-file (slurp "input.txt"))
(defn write-output [output]
  (with-open [output-file (clojure.java.io/writer "output.txt")]
    (doseq [line output]
      (println output-file line))))

(defn -main []
  (let [lines (clojure.string/split-lines input-file)
        t (Integer. (first lines))
        data (rest lines)
        output (atom [])]
    (loop [t t
           data data]
      (when (pos? t)
        (let [n (Integer. (first data))]
          (def graph (zipmap (range n) (repeat #{})))
          (def data (rest data))
          (loop [data data]
            (let [line (first data)]
              (if (empty? line)
                (do
                  (swap! output conj (kahn graph n))
                  (def data (rest data))
                  (recur data))
                (let [u v (map read-string (clojure.string/split line #" "))]
                  (swap! graph update-in [u] conj v)
                  (recur (rest data))))))))
        (recur (dec t) (rest data))))
    (write-output @output)))

(-main)
